lib/chef/resource/archive_file.rb

```diff
@@ -0,0 +1,135 @@
+#
+# Copyright:: Copyright 2017-2018, Chef Software Inc.
+# Author:: Jamie Winsor (<jamie@vialstudios.com>)
+# Author:: Tim Smith (<tsmith@chef.io>)
+#
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/resource"
+
+class Chef
+  class Resource
+    class ArchiveFile < Chef::Resource
+
+      resource_name :archive_file
+      provides :archive_file
+      provides :libarchive_file # legacy cookbook name
+
+      introduced "15.0"
+      description "Use the archive_file resource to extract archive files to disk. This resource uses the libarchive library to extract multiple archive formats including tar, gzip, bzip, and zip formats."
+
+      property :path, String,
+               name_property: true,
+               coerce: proc { |f| ::File.expand_path(f) },
+               description: "An optional property to set the file path to the archive to extract if it differs from the resource block's name."
+
+      property :owner, String,
+               description: "The owner of the extracted files"
+
+      property :group, String,
+               description: "The group of the extracted files"
+
+      property :mode, [String, Integer],
+               description: "The mode of the extracted files",
+               default: "755"
+
+      property :destination, String,
+               description: "The file path to extract the archive file to.",
+               required: true
+
+      property :options, [Array, Symbol],
+               description: "An array of symbols representing extraction flags. Example: :no_overwrite to prevent overwriting files on disk.",
+               default: lazy { [] }
+
+      # backwards compatibility for the legacy cookbook names
+      alias_method :extract_options, :options
+      alias_method :extract_to, :destination
+
+      action :extract do
+        description "Extract and archive file."
+
+        require "fileutils"
+
+        unless ::File.exist?(new_resource.path)
+          raise Errno::ENOENT, "No archive found at #{new_resource.path}!"
+        end
+
+        unless Dir.exist?(new_resource.destination)
+          converge_by("create directory #{new_resource.destination}") do
+            FileUtils.mkdir_p(new_resource.destination, mode: new_resource.mode.to_i)
+          end
+        end
+
+        converge_by("extract #{new_resource.path} to #{new_resource.destination}") do
```

Comment:
converge_by without a guard so this always fires and is always updated, isn't idempotent.

Reply:
At the moment this isn't actually idempotent and should really only be used with a notification block